java/feathub-udf/src/main/java/com/alibaba/feathub/udf/AbstractTimeWindowedAggFunc.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.flink.table.functions.AggregateFunction;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.TypeInference;
+import org.apache.flink.table.utils.DateTimeUtils;
+import org.apache.flink.util.Preconditions;
+
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Base class for all the time windowed aggregation function.
+ *
+ * <p>The aggregate function should be called with an timeInterval. It only aggregates rows with row
+ * time in the range of [rowTime - timeInterval, rowTime]. Currently, Flink SQL/Table only support
+ * over window with either time range or row count-based range. This can be used with a row
+ * count-based over window to achieve over window with both time range and row count-based range.
+ *
+ * @param <INPUT_T> The type of the input.
+ * @param <RESULT_T> The result type of the aggregation.
+ */
+public abstract class AbstractTimeWindowedAggFunc<INPUT_T, RESULT_T>
+        extends AggregateFunction<
+                RESULT_T, AbstractTimeWindowedAggFunc.TimeWindowedAccumulator<INPUT_T>> {
+
+    /** The time range of rows to include in the aggregation. */
+    protected Duration timeInterval;
+
+    @Override
+    public RESULT_T getValue(TimeWindowedAccumulator<INPUT_T> accumulator) {
+        final Long lowerBound = accumulator.latestTimestamp - timeInterval.toMillis();
+        List<INPUT_T> values = new LinkedList<>();
+        try {
+            for (Map.Entry<Long, List<INPUT_T>> entry : accumulator.values.entries()) {
+                if (entry.getKey() < lowerBound) {
+                    continue;
+                }
+                values.addAll(entry.getValue());
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        return getValue(values);
+    }
+
+    /**
+     * Compute the aggregation result with the given value.
+     *
+     * @param values The values that are in the time window range.
+     * @return The aggregation result.
+     */
+    protected abstract RESULT_T getValue(List<INPUT_T> values);
+
+    @Override
+    public TimeWindowedAccumulator<INPUT_T> createAccumulator() {
+        return new TimeWindowedAccumulator<>();
+    }
+
+    public void accumulate(
+            TimeWindowedAccumulator<INPUT_T> acc, Duration duration, INPUT_T value, Timestamp ts)
+            throws Exception {
+        if (this.timeInterval == null) {
+            this.timeInterval = duration;
+        } else {
+            Preconditions.checkState(
+                    this.timeInterval.equals(duration), "timeInterval should not changes.");
+        }
+        final long timestamp = DateTimeUtils.toInternal(ts);
+        acc.addValue(value, timestamp);
+    }
+
+    public void accumulate(
```

Comment:
This PR allows the AbstractTimeWindowedAggFunc to be called with 3 different timestamp types, i.e. `Instant`, `Timestamp` and `Instant`. Can FeatHub uses only one type consistently instead of three types, e.g. the type of `native_flink_expr.col(EVENT_TIME_ATTRIBUTE_NAME)`?